Addition of two new MANYBODY pair styles (sw/angle/table and threebody/table)

schererc opened the pull request:
As agreed to by Axel Kohlmeyer, I have created a pull request for two additional pair styles adding functionality for 3-body forces. As one is derived from the Stillinger-Weber pair style, I have added both into the folder src/MANYBODY.

The first pair style (sw/3b/table) is a modification of the Stillinger-Weber pair style with tabulated angular interactions for the three-body part. It is derived from it. It has been orignially developed for coarse-grained simulations (of water) and used in the following publication:

C. Scherer, and D. Andrienko, Understanding three-body contributions to coarse-grained force fields, Phys. Chem. Chem. Phys, 20(34):22387–22394, 2018, http://xlink.rsc.org/?DOI=C8CP00746B [<http://xlink.rsc.org/?DOI=C8CP00746B>](http://xlink.rsc.org/?DOI=C8CP00746B)

In this respect, I made the function "read_file(char *)" of pair_sw.h virtual to allow overriding by pair_sw_3b_table.h

The documentation is added as doc/src/pair_sw_3b_table.rst
An example is added as an extra folder to: examples/PACKAGES/pair_sw_3b_table


The second pair style (3b/table) is a pair style for generic fully tabulated three-body forces. It has been developed for (coarse-grained) simulations with Kernel-based machine learning (ML) potentials and used in the following publication:

C. Scherer, R. Scheid, D. Andrienko, and T. Bereau, Kernel-Based Machine Learning for Efficient Simulations of Molecular Liquids, J. Chem. Theor. Comp., 16(5):3194–3204, 2020, https://doi.org/10.1021/acs.jctc.9b01256 [<https://doi.org/10.1021/acs.jctc.9b01256>](https://doi.org/10.1021/acs.jctc.9b01256)

The documentation is added as doc/src/pair_3b_table.rst
An example is added as an extra folder to: examples/PACKAGES/pair_3b_table

I think, both pair styles could add extra functionality to LAMMPS in order to compute tabulated (generalized) 3-body forces.

I already thank you for reviewing my pull request. Feel free to contact me.

With Best Regards

Christoph Scherer

-- 
Dr. Christoph Scherer
Postdoc
Max Planck Institute for Polymer Research
Ackermannweg 10
D-55128 Mainz
Tel.: +49 6131 379 167
Email: scherer@mpip-mainz.mpg.de
           christophscherer83@gmail.com

akohlmey commented:
@schererc I added a bunch of changes for a better alignment with LAMMPS programming style and conventions (did *not* enable and apply clang-format) on the *.cpp files, though, and integration into the general "code management". Same for the documentation.

I am a bit confused by the differences in tables and tabulation. It looks like pair style sw/3b/table would be more correctly named as sw/angle/table as the table format and syntax in the pair style has close similarity to angle style table, while pair style 3b/table has a different format. The current naming would suggest that the tables would be interchangeable (at least their format).

I already commented on the lack of energy assignment for 3b/table. That is a significant deficiency.

The examples also need to be better aligned with our usual conventions, but before looking into that, I would like to hear everybody's comment on the name and the lack of energies for pair style 3b/table.

schererc commented on doc/src/pair_3b_table.rst:
```diff
@@ -0,0 +1,265 @@
+(of water) with Kernel-based machine learning (ML) potentials
+(:ref:`Scherer2 <Scherer2>`).  As for the pair style :doc:`pair_style sw
+<pair_sw>` or :doc:`pair_style sw/3b/table <pair_sw_3b_table>`, the energy of
+a system is computed as a sum over three-body terms:
```
What I saw is that I forgot to to transfer the read in energies from the table column. I added the corresponding line now into the function Pair3BTable::uf_lookup( ). Now, the issue with the uninitialized data should be solved. It is true that in the examples the energy column of the tables is zero. That was due to the parametrization, but the pair style should allow for correct energy treatment.

akohlmey replied:
> It is true that in the examples the energy column of the tables is zero. That was due to the parametrization, but the pair style should allow for correct energy treatment.
Would it be possible to change the example to one that *does* include the energy? Otherwise we should add a README file explaining it to the examples folder. Otherwise people might just use the table files for their own purposes and expect that there are meaningful energies and get inconsistent results (or worse).

schererc replied:
Sure, in principle yes. The issue is, honestly, that I/we generated the table with the VOTCA kernel based ML code, where we do generalized force matching and therefore at the moment only can calculate / output forces and not the energies (yet). We tested everything and the MD simulation with the forces works fine and can reproduce results of model systems (as the example) also without the energies. The idea was of course to include the energy column into the table and the pair style though it has not been used so far. Of course, one should comment on this explicitly in the README file! I can do this

schererc commented:
It is true, I should probably rename the pair sytle sw/3b/table to sw/angle/table, as the read in table corresponds to the style of angle style table and is not interchangeable with the table format of the pair style 3b/table. The idea of the latter was to have a way to read in generic 3-body forces. If agreed to, I could change everything from sw/3b/table to sw/angle/table?

akohlmey commented:
> It is true, I should probably rename the pair sytle sw/3b/table to sw/angle/table, as the read in table corresponds to the style of angle style table and is not interchangeable with the table format of the pair style 3b/table. The idea of the latter was to have a way to read in generic 3-body forces. If agreed to, I could change everything from sw/3b/table to sw/angle/table?

@sjplimp what do you think?

sjplimp commented:
I agree that sw/angle/table is a better name.  Also, I don't like 3b/table as a name.  If nothing else, it will make
it hard to find in the alphabetic listings of pair styles.  I don't think there are any styles in LAMMPS (pair or otherwise)
which start with a number.  Maybe threebody/table?

athomps commented:
> I agree that sw/angle/table is a better name. Also, I don't like 3b/table as a name. If nothing else, it will make
> it hard to find in the alphabetic listings of pair styles. I don't think there are any styles in LAMMPS (pair or otherwise)
> which start with a number. Maybe threebody/table?
I agree that 3b/table is a bad name.  threebody/table would not be very good either, because it is too generic. Looking at the doc page for `pair_3b_table.rst` it is clear that this potential is generated by VOTKA and/or msg_ml.  It would make sense to indicate that connection in the name e.g. pair_votka3b.rst (why say table when there is no non-table version?).

schererc commented:
Okay, sure pair_votca3b.rst would be also possible. However, I would prefer the proposed name threebody/table. Sure, there is no non-table version and the potential is generated by VOTCA. However, the idea was that there is a generalized tabulated potential for threebody forces that can also be generated in a different way (not by VOTCA). Therefore, why not keeping table in the name? That would be also closer to how we called the pair style in our original paper. However, that's my opinion.

If we agree on the (re-)naming of both styles, I could change/adapt the code?

Thanks a lot for the fast review!

akohlmey commented:
I would prefer threebody/table over votca/3b as well, since the latter would imply that we are either calling out to VOTCA or using some feature that is specific to VOTCA. But from the code it looks like it just provides a potential surface indexed by the three distances between the three interacting particles.

> That would be also closer to how we called the pair style in our original paper.

That is a risk you take if you publish the paper before integrating the code into the LAMMPS distribution. :wink:

akohlmey commented:
> If we agree on the (re-)naming of both styles, I could change/adapt the code?

Yes, please go ahead. If the pull request cannot be merged by tomorrow, it will likely miss the upcoming stable release.
We plan the next patch release for tomorrow and after that will only merge bugfixes for a couple of weeks to get to the stable release.

schererc commented:
> > If we agree on the (re-)naming of both styles, I could change/adapt the code?
> 
> Yes, please go ahead. If the pull request cannot be merged by tomorrow, it will likely miss the upcoming stable release. We plan the next patch release for tomorrow and after that will only merge bugfixes for a couple of weeks to get to the stable release.

Thanks! Now with one day delay, I have changed everything according to the name change(s) including examples and documentation files. Please let me know if it is okay like this!

schererc commented:
> > Thanks! Now with one day delay, I have changed everything according to the name change(s) including examples and documentation files. Please let me know if it is okay like this!
> 
> There were some minor changes still required for the documentation to be consistent and according to the conventions we use.
> 
> Also, I found and fixed a memory access issue when running in parallel and protected the table and parameter reading from leaking memory.
> 
> Next, I merged the two example folders (since it is essentially 3x the same system) and updated/simplified the examples according to our conventions and also streamlined the README file. For a real tutorial, it was explaining too little (and then it should be a Howto document in the manual) and for an example too much. The purpose of examples is just to show the features in use, not to reproduce some published results. You are welcome to contribute a Howto or let us add a URL to a website you maintain with a more detailed tutorial according to your taste.
> 
> Finally, I've added yaml files for our unit test tools. While checking those, I notices that the differences in pressure computed via F dot r (the "normal" way) and with pair_modify nofdotr (via ev_tally*()) is not consistent for the `threebody/table` style. It is not a large difference, but large enough to be a little bit uncomfortable about it. Pair style sw/angle/table does not have this issue.
> 
> Example for the `in.spce` the first pressure is `-320.10184` versus `-320.42417` and it continues to have a difference of similar magnitude (~0.35 pressure units) throughout, which forces and energies are - of course - identical.
> 
> I am ok to merge this "as is", as this nofdotr check is a rather minor issue, but we should make an effort to figure out what causes this within the next couple of weeks.
> 
> @sjplimp @athomps any objections? or alternate suggestions?

Thanks a lot!

Regarding (extended) tutorials, you can of course link to: https://gitlab.mpcdf.mpg.de/votca/votca/-/tree/master/csg-tutorials

Or explicitly for the pair style sw/angle/table to: https://gitlab.mpcdf.mpg.de/votca/votca/-/tree/master/csg-tutorials/spce/3body_sw

And for the pair style threebody/table to: https://gitlab.mpcdf.mpg.de/votca/votca/-/tree/master/csg-tutorials/ml

akohlmey commented:
> Or explicitly for the pair style sw/angle/table to: https://gitlab.mpcdf.mpg.de/votca/votca/-/tree/master/csg-tutorials/spce/3body_sw
> 
> And for the pair style threebody/table to: https://gitlab.mpcdf.mpg.de/votca/votca/-/tree/master/csg-tutorials/ml

Just added those URLs and also put a reference to the example inputs at the top (as suggested in issue #2680 )

athomps reviewed (approved):
I approve.
